py/farm_ng/core/event_service.py

```diff
@@ -0,0 +1,445 @@
+            self._client_queues[request.uri.path] = []
+
+        self._client_queues[request.uri.path].append(request_queue)
+
+        def delete_queue(_) -> None:
+            """Callback to delete the queue when the client disconnects."""
+            # remove the queue from the list of queues
+            self._client_queues[request.uri.path].remove(request_queue)
+
+            # check that there are no more clients connected
+            self.update_state_idle()
+
+        # register the callback to delete the queue when the client disconnects
+        context.add_done_callback(delete_queue)
+
+        while True:
+            yield await request_queue[1].get()
+
+    def _send_event_payload(self, event: Event, payload: bytes) -> None:
+        """Send an event and payload to the clients.
+
+        Args:
+            event (Event): The event.
+            payload (bytes): The payload.
+        """
+        uri: Uri = event.uri
+
+        # register the URI of the message
+        self._uris[uri.path] = uri
+
+        reply = SubscribeReply(
+            event=event,
+            payload=payload,
+        )
+
+        # get client queues for the URI
+        client_queues: list[tuple[SubscribeRequest, asyncio.Queue]]
+        client_queues = self._client_queues.get(uri.path, [])
+        client_queues = self._client_queues.get("*", []) + client_queues
+
+        queues: list[asyncio.Queue] = [
+            queue
+            for request, queue in client_queues
+            if request.every_n == 0 or reply.event.sequence % request.every_n == 0
+        ]
+        self.logger.debug(
+            f"Sending {uri.path}: {reply.event.sequence} to {len(queues)} clients"
+        )
+
+        # send the reply to the clients
+        for queue in queues:
+            try:
+                queue.put_nowait(reply)
+            except asyncio.QueueFull:
+                # NOTE: don't we want to drop the oldest message?
+                pass
+
+    def _send_raw(
+        self, uri: Uri, message: Message, timestamps: list[Timestamp]
+    ) -> None:
+        """Send a message to the service.
+
+        Args:
+            uri (Uri): The URI of the message.
+            message (Message): The message.
+            timestamps (list[Timestamp]): The timestamps.
+        """
+        # register the URI of the message
+        self._uris[uri.path] = uri
+
+        # get the count of the URI, or 0 if it doesn't exist
+        count = self._counts.get(uri.path, 0)
+
+        # increment the count of the URI
+        self._counts[uri.path] = count + 1
+
+        # create the event and send
+        payload: bytes = message.SerializeToString()
+
+        event = Event(
+            uri=uri,
+            timestamps=timestamps,
+            payload_length=len(payload),
+            sequence=count,
+        )
+        self._send_event_payload(event, payload)
+
+    # TODO: i'd rename this to `publish` to match with ROS
```

Comment:
This isn't really publish, it allows a client to send a message directly to a service, without the service establishing a subscription. Publish is implemented by sending messages to all the subscribers.

We don't have a centralized broker here. The alternative is to remove this function and declare all subscriptions in the config file, but that seems complex and doesn't allow a simple client to come up and start sending messages directly to a service.

Note these messages currently wouldn't be logged as they are point to point, was investigating republishing them.

This specific interface is still wip for sure.

Reply:
Do we want/need a centralised broker? That might solve the issue to pass through messages over different services

Reply:
I would prefer not to implement a broker yet, I think a config file will be sufficient for a while.

Reply:
I renamed this function to reqRep (request reply).

Reply:
i like more explicit names -> `requestReply`